src/crimson/net/Workspace.h

```diff
@@ -0,0 +1,59 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2018 Intel Corp.
+ *
+ * Author: Yingxin Cheng <yingxincheng@gmail.com>
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#pragma once
+
+#include "Protocols.h"
+
+namespace ceph::net {
+
+class Workspace
+{
+  NoneProtocol none_protocol;
+  AcceptProtocol accept_protocol;
+  ConnectProtocol connect_protocol;
+  OpenProtocol open_protocol;
+  CloseProtocol close_protocol;
+  Protocol *const protocols[static_cast<int>(state_t::LAST)] = {
```

Comment:
i don't think we need to index the states with an array. the next state to transit to is determined at compile time.

Reply:
I think this needs to modify class dependencies to expose `Workspace` to concrete `Protocol` classes, will try.